src/core/thread/network_diagnostic.cpp

```diff
@@ -561,14 +548,8 @@ Error NetworkDiagnostic::SendDiagnosticReset(const Ip6::Address &aDestination,
     Coap::Message *  message = nullptr;
     Tmf::MessageInfo messageInfo(GetInstance());
 
-    VerifyOrExit((message = Get<Tmf::Agent>().NewMessage()) != nullptr, error = kErrorNoBufs);
-
-    SuccessOrExit(error = message->InitAsConfirmablePost(UriPath::kDiagnosticReset));
-
-    if (aCount > 0)
```

Comment:
The `aCount > 0` check is removed. 
Is it intentional ?

Reply:
Yes. The check for adding payload marker is redundant and not needed.

I mentioned this in the documentation of the methods.
```c++
    /**
     * This method allocates and initializes a new CoAP Confirmable Post message with Network Control priority level.
     *
     * The CoAP header is initialized as `kTypeConfirmable` and `kCodePost` with a given URI path and a randomly
     * generated token (of default length). This method also sets the payload marker (`SetPayloadMarker()` on message).
     * Even if message has no payload, calling `SetPayloadMarker()` is harmless, since `SendMessage()` will check and
     * remove the payload marker when there is no payload.
     *
     * @param[in] aUriPath   The URI path string.
     *
     * @returns A pointer to the message or `nullptr` if failed to allocate message.
     *
     */
```

Reply:
Here is the code related to this in `coap_message.hpp`. `Finish()` is always called before we send a message (it writes the CoAP header into the message).

```c++
void Message::Finish(void)
{
    // If the payload marker is set but the message contains no
    // payload, we remove the payload marker from the message. Note
    // that the presence of a marker followed by a zero-length payload
    // will be processed as a message format error on the receiver.

    if (GetHelpData().mPayloadMarkerSet && (GetHelpData().mHeaderLength == GetLength()))
    {
        IgnoreError(SetLength(GetLength() - 1));
    }

    WriteBytes(0, &GetHelpData().mHeader, GetOptionStart());
}
```